### PR TITLE
 tests: Add Python dead code linter (vulture) to Travis

### PR DESCRIPTION
**Pull request by practicalswift**
Add Python dead code linter (`vulture`) to Travis.

Rationale for allowing dead code only after explicit opt-in (via `--ignore-names`):
* Less is more :-)
* Unused code is by definition "untested"
* Unused code can be an indication of bugs/logical errors. By making the contributor aware of newly introduced unused code it gives him/her an opportunity to investigate if the unused code they introduce is malignant or benign :-)
* Unused code is hard to spot for humans and is thus often missed during manual review
* [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it)

Based on #14312 to make linter job pass.

### CODE REVIEW

**Review by promag — COMMENTED** [review]
In my system it takes around 1 second so it's not a big deal.

I suggest closing #14312.

**Comment by practicalswift** [comment]
@promag Thanks for reviewing. Good idea! Done.

File: test/functional/test_framework/test_framework.py [hunk 1]
```diff
@@ -161,8 +161,10 @@ def main(self):
         success = TestStatus.FAILED
 
         try:
-            if self.options.usecli and not self.supports_cli:
-                raise SkipTest("--usecli specified but test does not support using CLI")
+            if self.options.usecli:
+                if not self.supports_cli:
+                    raise SkipTest("--usecli specified but test does not support using CLI")
+                self.skip_if_no_cli()
```
Comment (MarcoFalke) [inline test/functional/test_framework/test_framework.py]: This hunk probably needs testing. Mind to submit as separate pull request?

**Comment by practicalswift** [comment]
@MarcoFalke Updated. Please re-review :-)

File: test/functional/test_framework/messages.py [hunk 2]
```diff
@@ -893,7 +889,6 @@ def __init__(self):
         self.nTransactions = 0
         self.vHash = []
         self.vBits = []
-        self.fBad = False
```
Comment (jb55) [inline test/functional/test_framework/messages.py]: even though this is unused, is it really a good idea to remove code that initializes default values?

Comment (practicalswift) [reply test/functional/test_framework/messages.py]: @jb55 Do you mean the fact that `fBad` was still in `__slots__`? Now it is gone so `fBad` is history. Updated version looks good? :-)

Comment (jb55) [reply test/functional/test_framework/messages.py]: perhaps, 3a4449e9ad945313c6637283757de8d539cf790f states:

> Classes use __slots__ to ensure extraneous attributes aren't accidentally added by tests, compromising their intended effect.

this seems to be reversing the original intent of that commit. /cc @JustinTArthur ?

Comment (JustinTArthur) [reply test/functional/test_framework/messages.py]: If `fBad` isn't going to be used, removing it won't go against the intentions of the slots work, which is just there to protect test developers from making tests that don't do what the developer thinks it does.

I just want to make sure the removal doesn't break the intentions of those message module classes, which provide test developers with Python equivalents of the common Core data objects—`fBad` is still in the Core C++ classes. If we generally only include attributes we need for existing tests, then no problem removing it and someone can re-add it later if their test incorporates it. I don't have enough history with the functional test suites to make a judgement call.

Comment (promag) [reply test/functional/test_framework/messages.py]: +1 remove `fBad`.

**Comment by promag** [comment]
Should explicitly set `--min-confidence`? Currently it's [60](
https://github.com/jendrikseipp/vulture/blob/62c379de033bfdb690d0ecc835b64fc8c3dddc6c/vulture/core.py#L43).

File: test/functional/test_framework/messages.py (hunk 2 above)
Comment (jb55) [reply test/functional/test_framework/messages.py]: +0 sounds good to me.

nit: it looks like the fBad slot removal was squashed into the wrong commit.

Comment (practicalswift) [reply test/functional/test_framework/messages.py]: @jb55 Thanks for letting me know about the squash mixup. Now fixed :-)

**Comment by practicalswift** [comment]
@promag Judging from my testing of `vulture` over the set of the 196 mergeable open PR:s the default of 60 seems reasonable. Should we keep the default for now?

**Comment by promag** [comment]
My suggestion is to add ` --min-confidence 60`.

Also, from the vulture documentation:

> We recommend using whitelists instead of --ignore-names or --ignore-decorators whenever possible

WDYT?

**Comment by practicalswift** [comment]
@promag I've not added an explicit `--min-confidence 60`.

Regarding the white-list method: I'm afraid a whitelist replicating the current wildcard based method would be quite long. And perhaps burdensome to keep updated.

The trade-off looks like this:

```
    --ignore-names "argtypes,connection_lost,connection_made,converter,data_received,\
        daemon,errcheck,msg_generic,on_*,optionxform,restype,skip_if_no_cli"
```

Versus:

```
$ vulture --min-confidence 60 --make-whitelist $(git ls-files -- "*.py" ":(exclude)contrib/") | wc -l
110
```

Despite the recommendation I think `--ignore-names` might be the better option for us since we need the wildcard functionality.

File: test/lint/lint-python-dead-code.sh [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+export LC_ALL=C
+
+if ! command -v vulture > /dev/null; then
+    echo "Skipping Python dead code linting since vulture is not installed. Install by running \"pip3 install vulture\""
+    exit 0
+fi
+
+vulture \
+    --min-confidence 60 \
+    --ignore-names "argtypes,connection_lost,connection_made,converter,data_received,daemon,errcheck,msg_generic,on_*,optionxform,restype,skip_if_no_cli" \
```
Comment (MarcoFalke) [inline test/lint/lint-python-dead-code.sh]: why is `skip_if_no_cli` ignored?

**Comment by promag** [comment]
> @promag I've not added an explicit `--min-confidence 60`.

You have.

**Comment by practicalswift** [comment]
> > @promag I've not added an explicit `--min-confidence 60`.
> 
> You have.

Oh, "not" should have been "now" :-D

File: test/lint/lint-python-dead-code.sh (hunk 3 above)
Comment (practicalswift) [reply test/lint/lint-python-dead-code.sh]: Nice catch! Now removed!

**Comment by ken2812221** [comment]
utACK c422cf2b50b08b489172de28f69083784c35a455

**Comment by jb55** [comment]
utACK c422cf2b50b08b489172de28f69083784c35a455

**Comment by conscott** [comment]
utACk - vulture is great!

**Comment by practicalswift** [comment]
Added a commit which pins the `vulture` version to avoid the possibility of having the build broken due to a new `vulture` version being released.

Please re-review :-)

**Comment by practicalswift** [comment]
Rebased! :-)

File: test/functional/test_framework/key.py [hunk 4]
```diff
@@ -111,40 +110,12 @@ def set_secretbytes(self, secret):
...
-                                 ssl.EC_KEY_get0_public_key(other_pubkey.k),
-                                 self.k, 0)
-        if r != 32:
-            raise Exception('CKey.get_ecdh_key(): ECDH_compute_key() failed')
-        return ecdh_keybuffer.raw
-
-    def get_ecdh_key(self, other_pubkey, kdf=lambda k: hashlib.sha256(k).digest()):
-        # FIXME: be warned it's not clear what the kdf should be as a default
-        r = self.get_raw_ecdh_key(other_pubkey)
-        return kdf(r)
```
Comment (MarcoFalke) [inline test/functional/test_framework/key.py]: This is a library (ECC secp256k1 OpenSSL wrapper), so removing the wrapper functions might break code that is not in this commit. (Maybe a different branch or repo, and it will break on merge or rebase)

Comment (practicalswift) [reply test/functional/test_framework/key.py]: Good point! Updated. No longer removing these functions. Please re-review :-)

**Comment by MarcoFalke** [comment]
Going to merge this since it had 3 utACKs and is probably not worth to wait for all re-ACKs.
